### DISCUSSION TITLE
using fakeSMTP's along with Laragon

### DISCUSSION THREAD

**Discussion by orionseye** [leokhoa/laragon#733] [General]
Hello good people,

i am having a hard time to setup any fake smtp server (for testing) while using Laragon.
I don't want to use the build in mail sender since i am not interested emails to be delivered.
So i tried almost all fake SMTP servers out there, from smtp4devs, papercut .. and  a bunch of others.
None of them seem to be able to listen to port 25 and catch my mails.

The only workaround i found is to enable both mail catcher & mail sender in Laragon and insert some fake gmail user/password to get at least the mails sent in my outputs folder.

Has anyone experienced the same issue?
Can it be that due that Laragon reserves port 25 for catching & delivery, other apps won't be able to catch mails from port 25?

WIN 10
Laragon Mint 5.0.0 210523
PHP 7.4
Apache24

**Comment by maulanamukarom** [comment]
try this: https://github.com/ChangemakerStudios/Papercut-SMTP

**Comment by ppfoong** [comment]
Why not using the real gmail SMTP? It is pretty easy to set up.

You can also use outlook.com SMTP too. I assume you have one, as nowadays Microsoft is forcing Windows users to have one.

Both gmail and outlook.com SMTP set up are similar. You need to go to your account setup page, and setup an application password, then use that application password as your SMTP sender's password.

**Reply by orionseye** [reply]
As described, I tried that but did not work. Other fake SMTP apps did not work as well.
The issue was, that (after debugging) I found an error message notifying "Warning: mail(): Bad Message Return Path "
The fix was to uncomment in php.ini "sendmail_from = me@example.com"  .. and this fixed the issue

**Reply by Andre-ADPC** [reply]
Gmail's **"Less secure apps"** Access has been giving issues for a while on my setup, and Google has put an end to this for good now as well. Too many CySec issues with it, so we need to find alternative solutions. [Reference Google](https://support.google.com/accounts/answer/6010255?hl=en).

In my case Laragon's **Mail Analyzer** returns all 220 and 25 SMTP server codes until a final "535-5.7.8 Username and Password not accepted.", closing the connection with 221.

The Gmail account used is valid, active and working. But I assume from the Laragon **Mail Analyzer** log feedback that Google's Gmail servers are throwing the final auth request out, as it's deemed insecure & invalid.

**PaperCut SMTP** seemed promising, but it's not working for Laragon seeing it requires no authentication layer and can not be configured via the Laragon UI. I had a look at all the instances of **`gmail.smtp`** in the Laragon Directory, doing a search using my IDE, but apart from the  **`sendmail.ini`** file, hinting there could be some configuration link to the Laragon UI, nothing else but the **` *.phps`** config files, Python's email config, and my MINGW64 info page's HTML file came up in the search.  

"Hacking the **`sendmail.ini`** in an attempt to crowbar **PaperCut SMTP** to work in the place of **Gmail** was fruitless.

Changing the _**php.ini**_ file as suggested by @orionseye above, does not deliver any results in my case either.

### System
WIN 10
Laragon Full 6.0.0 220916
PHP 8.2 TS
Apache 2.4.52 8080/XXX
Nginx 1.21.4   8089/XXX

What options do we have or maybe some references to a detailed guide we can follow to resolve the issue? 
Any tips or pointers will be highly appreciated.

**Reply by ppfoong** [reply]
You just need to setup an "app password" in Google account settings, then use that specific app password for your SMTP account.

Use this link to go to the Google App Password setup page:
https://myaccount.google.com/apppasswords

**Reply by Andre-ADPC** [reply]
Thank you for your reply @ppfoong 
Google throws a tantrum, stating "The setting you are looking for is not available for your account." 

Why, I have no clue. It's a straightforward `xyz@gmail.com` account. I have been battling with this for long enough. Time to move on.

Relying on other SMTP vendors like Google, MS, etc is not secure anyway, seeing that during dev cycles some data that's not supposed to be flung into the cloud might get sent to email boxes, and email is NOT secure at all. Proven over and over, so scrapping the idea of using the mail functionality of Laragon for now.

**Comment by orionseye** [comment]
@Andre-ADPC  i am quite sure this is  due to your php settings, but I might be wrong.
It makes sense to try some things out:

- Read the PHP mail configuration  https://www.php.net/manual/en/mail.configuration.php
- Open CMD and type `telnet localhost 25` check the results
- Make sure, in your php.ini SMTP is 'localhost' & smtp_port is 25:
[mail function]
; For Win32 only.
; http://php.net/smtp
SMTP = localhost
; http://php.net/smtp-port
smtp_port = 25
- create a email_test.php file and dump these line inside
```
ini_set( 'display_errors', 1 );
error_reporting( E_ALL );
$sender = 'someone@somedomain.tld';
$recipient = 'you@yourdomain.tld';

$subject = "php mail test";
$message = "php test message";
$headers = 'From:' . $sender;

if (mail($recipient, $subject, $message, $headers))
{
    echo "Message accepted";
}
else
{
    echo "Error: Message not accepted";
}
```

With the above, and display errors on, you might catch any PHP related issue.
If you get a Warning: mail(): Bad Message Return Path .. it's definitely your php.ini setting

We are here to help

**Reply by Andre-ADPC** [reply]
Hi @orionseye ,
Thanks for the tips. My `php.ini` file is 100%. `telnet`, `Test-NetConnection 127.0.0.1 -p 25`, and even `nmap -p 25 127.0.0.1` ran from my WSL2 reports all as expected. My last config was for PaperCut SMTP, and that's what's reported via Linux, MINGW64 and Win PS. No issues with the `php.ini` file.

Laragon's "backend" has Google's SMTP server locked in, seeing no matter what email address & creds you try, the Laragon Mail Analyzer spits out a "Google `smtp.gmail`" error.

**Comment by Pen-y-Fan** [comment]
I use Mailpit to test emails locally. I have a blog [How to install mailpit on Windows and integrate with Laragon](https://pen-y-fan.github.io/2023/02/23/how-to-install-mailpit-in-laragon/). I hope this helps.

**Reply by Andre-ADPC** [reply]
Hi @Pen-y-Fan , thank you for the pointer. I was planning on having a look at MailPit and comparing the benefits against PaperCut. It sure seems that it could do the trick. Lots of practical functionality in there. Cypress integration testing is also a great feature to have, so it will be the go-to for me going forward. 

The Laragon email tools used to work on my system, and were nice to use as a backup check, but I'm disabling them seeing it's not wanting to get gonfigured, LOL. When things "break", it always raises some "alarms" and concerns that I have messed up somewhere by installing whatever... Win update, new this-that-or the other. Just to find out later, no, it wasn't your doing. 

There are so many constantly changing variables involved,  and then also the juggling in the background with the `*.ini` and `xyz.php` config files on top of general OS services; AV software, firewalls, VPNs, etc. So to try and figure this Laragon situation out for what you currently get is not worth it IMO.

**Reply by Andre-ADPC** [reply]
Oh, I wanted to add:
Your newer Blog and previous contributions on the old Laragon forum is and always have been pure **GOLD**. Thank you for that and the efforts you put into them, Michael.
